### PR TITLE
Adding Cloudy tendencies

### PR DESCRIPTION
**Pull request by edejong-caltech**
## Content

----
- [ ] I have read and checked the items on the review checklist.

### CODE REVIEW

File: src/Common/initial_condition.jl [hunk 1]
```diff
@@ -198,6 +198,52 @@ function initial_condition_1d(
...
+    #     nexp = 2
+    # end
+    # pdists::NTuple{ND, <: CL.ParticleDistributions.PrimitiveParticleDistribution{FT}} = ntuple(ND) do k
+    #     if k <= nexp
+    #         CL.ParticleDistributions.ExponentialPrimitiveParticleDistribution(FT(0), FT(k))
+    #     else
+    #         CL.ParticleDistributions.GammaPrimitiveParticleDistribution(FT(0), FT(k), FT(1))
+    #     end
+    # end
+    pdists::NTuple{Int(NM / 3), CL.ParticleDistributions.GammaPrimitiveParticleDistribution{FT}} = ntuple(Int(NM/3)) do k
```
Comment (edejong-caltech) [inline src/Common/initial_condition.jl:237]: Move this to initialisation in aux, rather than in the state vector

File: src/Common/ode_utils.jl [hunk 2]
```diff
@@ -71,13 +71,28 @@ function initialise_state(::NonEquilibriumMoisture, ::Precipitation2M, initial_p
         N_aer = initial_profiles.N_aer,
     )
 end
+function initialise_state(::NonEquilibriumMoisture, ::PrecipitationNM, initial_profiles)
+    return CC.Fields.FieldVector(;
+        ρq_tot = initial_profiles.ρq_tot,
```
Comment (edejong-caltech) [inline src/Common/ode_utils.jl:76]: Eliminate most of these variables and just track moments

File: src/Common/ode_utils.jl [hunk 3]
```diff
@@ -133,4 +148,13 @@ function initialise_aux(FT, ip, common_params, thermo_params, air_params, activa
         Stats = Stats,
         TS = TS,
     )
+    if psNM
```
Comment (edejong-caltech) [inline src/Common/ode_utils.jl:151]: Initialise the pdists in here

File: src/Common/tendency.jl [hunk 4]
```diff
@@ -415,6 +424,69 @@ end
...
+    q_tot,
+    q_liq,
+    moments,
+    old_pdists,
+    T,
+    ρ,
+)
+    FT = eltype(q_tot)
+
+    # TODO: store/compute these only once
```
Comment (edejong-caltech) [inline src/Common/tendency.jl:441]: Do this in aux instead

File: src/Common/tendency.jl [hunk 5]
```diff
@@ -415,6 +424,69 @@ end
...
+    S_moments = ntuple(_ -> FT(0), length(moments))
+
+    q = TD.PhasePartition(q_tot, q_liq, FT(0))
+    
+    # update the ParticleDistributions
+    mom_norms = CL.get_moments_normalizing_factors(NProgMoms, norms)
+    mom_normed = tuple(moments ./ mom_norms...)
+    pdists = ntuple(length(old_pdists)) do i
+        ind_rng = CL.get_dist_moments_ind_range(NProgMoms, i)
+        CL.ParticleDistributions.update_dist_from_moments(old_pdists[i], mom_normed[ind_rng])
```
Comment (edejong-caltech) [inline src/Common/tendency.jl:456]: Add diagnosis of q_tot, q_liq in the update_aux_thermo

File: src/Common/ode_utils.jl (hunk 3 above)
Comment (edejong-caltech) [reply src/Common/ode_utils.jl:151]: Also add a struct for `cloudy_params` here to store all of the velocity, kernels, etc. stuff -- make them

File: src/Common/tendency.jl [hunk 6]
```diff
@@ -658,3 +752,17 @@ end
...
+
+    @. dY.ρq_rai += aux.moisture_variables.ρ * aux.precip_sources.S_q_rai
+    @. dY.N_liq += aux.precip_sources.S_N_liq
+    @. dY.N_rai += aux.precip_sources.S_N_rai
+
+    @. dY.N_liq += aux.activation_sources.S_N_liq
+    @. dY.N_aer += aux.activation_sources.S_N_aer
+    # TODO: add the activation tendency to the moments as well
+
+    @. dY.moments = aux.cloudy_sources.S_moments
```
Comment (edejong-caltech) [inline src/Common/tendency.jl:765]: Create an ntuple of zeros and put S_activation as the first value

Comment (edejong-caltech) [reply src/Common/tendency.jl:765]: Or worst case, keep the first distribution separate

File: test/experiments/KiD_driver/parse_commandline.jl [hunk 7]
```diff
@@ -13,15 +13,11 @@ function parse_commandline()
         arg_type = String
         default = "Float64"
         "--moisture_choice"
-        help = "Mositure model choice: EquilibriumMoisture, NonEquilibriumMoisture"
+        help = "Mositure model choice: EquilibriumMoisture, NonEquilibriumMoisture, CloudyMoisture"
         arg_type = String
         default = "NonEquilibriumMoisture"
-        "--prognostic_vars"
```
Comment (trontrytel) [inline test/experiments/KiD_driver/parse_commandline.jl:19]: I'm very happy to commit to one prognostic variable choice. Do you have a preference for theta or T? We should delete the other from CI and tendency options too

Comment (sajjadazimi) [reply test/experiments/KiD_driver/parse_commandline.jl:19]: I prefer T, because in PySDM KiD and also in Shipway and Hill T is fixed.

Comment (edejong-caltech) [reply test/experiments/KiD_driver/parse_commandline.jl:19]: Agreed on T

**Review by edejong-caltech — COMMENTED** [review]
Sajjad, this all looks fantastic -- thank you for merging my commits with the new KiD structure!

File: test/create_parameters.jl [hunk 8]
```diff
@@ -85,4 +86,36 @@ function create_kid_parameters(toml_dict)
     return kid_params
 end
 Base.broadcastable(x::KID.Parameters.KinematicDriverParameters) = Ref(x)
+
+function create_cloudy_parameters(NM, FT)
```
Comment (trontrytel) [inline test/create_parameters.jl:90]: Could you add some docstring explaining what is happening here? Or link to Cloudy documentation if it's explained there? I'm asking about `NM / 3`, `1e6` and `1e-9` in `norms`, `5e0` in `kernel_func`, `5e-10` `1e-6` in `KernelTensors` and `CoalescenceData` and `vel`, etc.

File: src/Common/tendency.jl [hunk 9]
```diff
@@ -155,6 +172,48 @@ end
...
+    (; weighted_vt) = aux.velocities
+
+    rain_number_ind = Int(CL.get_dist_moments_ind_range(cloudy_params.NProgMoms, 2)[1])
+    rain_mass_ind = Int(CL.get_dist_moments_ind_range(cloudy_params.NProgMoms, 2)[2])
+    @. N_liq = Y.moments.:1
+    @. N_rai = Y.moments.:($$rain_number_ind)
+    @. q_rai = q_(Y.moments.:($$rain_mass_ind), ρ)
+    @. moments = Y.moments
+
+    tmp = @. cloudy_precompute_aux_precip_helper(cloudy_params, moments, pdists)
```
Comment (trontrytel) [inline src/Common/tendency.jl:212]: I'm not very good at this, but I think that the way the Cloudy tendencies are added, both here and in 1D model, is allocating. Maybe instead we could create additional variables that we need for Cloudy and put them in `scratch` for temporary things that need to be stored during computations (like supersaturation, maybe `pdists`?) and in `aux` for the ones that need to be precomputed for tendencies. 

I was trying to write all the other tendencies like

```
@inline function precompute_aux_important_sources!
    source_term = aux.scratch.tmp
    (; stored_variable) = aux.microph_variables
 
    @. source_term = some_micro_foo(stored_variable)
    @. aux.important_sources = to_sources(-source_term, source_term)
end

@inline function important_sources_tendency!

    precompute_aux_important_sources!
    
    @. dY.ρq_tot += aux.thermo_variables.ρ * aux.important_sources.q_tot
end
```

It is more complicated for Cloudy than the other microphysics models. There are software office hours on Wednesday. If we can't figure it out by then, maybe we could submit it as a request to help then?

**Comment by sajjadazimi** [comment]
> Sajjad, this all looks fantastic -- thank you for merging my commits with the new KiD structure!

Thank you for adding Cloudy to KiD!

File: src/K1DModel/tendency.jl [hunk 10]
```diff
@@ -80,6 +80,89 @@ end
...
+
+    q = TD.PhasePartition(q_tot, q_liq, FT(0))
+    S::FT = TD.supersaturation(thermo_params, q, ρ, T, TD.Liquid())
+
+    (; r_dry, std_dry, κ) = kid_params
+    w = ρw / ρ
+
+    aerosol_distribution =
+        CMAM.AerosolDistribution((CMAM.Mode_κ(r_dry, std_dry, N_aer, (FT(1),), (FT(1),), (FT(0),), (κ,)),))
+    N_act = CMAA.N_activated_per_mode(activation_params, aerosol_distribution, air_params, thermo_params, T, p, w, q)[1]
```
Comment (trontrytel) [inline src/K1DModel/tendency.jl:111]: This is out of date with the aerosol activation PR that I merged, apologies!

Comment (trontrytel) [reply src/K1DModel/tendency.jl:115]: Can we use this one: `CMAA.total_N_activated` ? Or does it have to be only one aerosol mode for Cloudy right now?

**Pull request merged**